api/src/main/java/org/openmrs/module/reporting/dataset/definition/evaluator/IterableSqlDataSetEvaluator.java

```diff
@@ -0,0 +1,135 @@
+import org.openmrs.module.reporting.evaluation.EvaluationContext;
+import org.openmrs.module.reporting.evaluation.EvaluationException;
+import org.openmrs.module.reporting.common.SqlIterator;
+import org.openmrs.module.reporting.evaluation.querybuilder.SqlQueryBuilder;
+import org.openmrs.util.OpenmrsUtil;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.InputStream;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.util.Iterator;
+import java.util.Properties;
+
+/**
+ * The logic that evaluates a {@link IterableSqlDataSetDefinition} and produces an {@link DataSet}
+ *
+ * @see IterableSqlDataSetDefinition
+ */
+@Handler(supports = {IterableSqlDataSetDefinition.class})
+public class IterableSqlDataSetEvaluator implements IterableDataSetEvaluator {
+
+    private static final Logger log = LoggerFactory.getLogger(IterableSqlDataSetEvaluator.class);
+
+    /**
+     * Public constructor
+     */
+    public IterableSqlDataSetEvaluator() {
+    }
+
+    /**
+     * @should evaluate a IterableSqlDataSetDefinition
+     * @should evaluate a IterableSqlDataSetDefinition with parameters
+     * @should evaluate a IterableSqlDataSetDefinition with in statement
+     * @should protect SQL Query Against database modifications
+     * @see DataSetEvaluator#evaluate(DataSetDefinition, EvaluationContext)
+     */
+    public DataSet evaluate(DataSetDefinition dataSetDefinition, EvaluationContext context) throws EvaluationException {
+
+        context = ObjectUtil.nvl(context, new EvaluationContext());
+
+        IterableSqlDataSetDefinition sqlDsd = (IterableSqlDataSetDefinition) dataSetDefinition;
+
+        SqlQueryBuilder queryBuilder = new SqlQueryBuilder();
+
+        String sqlQuery = sqlDsd.getSql();
+
+        queryBuilder.append(sqlQuery);
+
+        queryBuilder.setParameters(context.getParameterValues());
+
+        IterableSqlDataSetDefinition defenition = (IterableSqlDataSetDefinition) dataSetDefinition;
+        Properties connectionProperties = getConnectionProperties(defenition.getConnectionPropertyFile());
+        Iterator iterator = null;
+        Connection connection = null;
+        try {
+            connection = createConnection(connectionProperties);
+            SqlRunner runner = new SqlRunner(connection);
+
+            if (StringUtils.isNotBlank(defenition.getSql())) {
+                iterator = runner.executeSqlToIterator(defenition.getSql());
+
+            }
+
+        } catch (EvaluationException ee) {
+            throw ee;
+        } catch (Exception e) {
+            throw new EvaluationException("An error occurred while evaluating a SqlFileDataSetDefinition", e);
+        }
+
+        return new IterableSqlDataSet(context, sqlDsd, (SqlIterator) iterator);
+    }
+
+    /**
+     * @return a new connection given a set of connection properties
+     */
+    protected Connection createConnection(Properties connectionProperties) throws EvaluationException {
```

Comment:
It might be better for this class to extend SqlFileDataSetEvaluator, and then evolve the methods in that parent class rather than copy and maintain a separate set of implementations of these methods.

Reply:
I do think that re-using logic makes sense.  I'd be more inclined to just adjust the SqlFileDataSetEvaluator such that the reusable pieces and customizable pieces are in appropriate methods that will optimize reusability.

Reply:
Done.